sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamTableFunctionScanRel.java

```diff
@@ -0,0 +1,157 @@
+  public BeamTableFunctionScanRel(
+      RelOptCluster cluster,
+      RelTraitSet traitSet,
+      List<RelNode> inputs,
+      RexNode rexCall,
+      Type elementType,
+      RelDataType rowType,
+      Set<RelColumnMapping> columnMappings) {
+    super(cluster, traitSet, inputs, rexCall, elementType, rowType, columnMappings);
+  }
+
+  @Override
+  public TableFunctionScan copy(
+      RelTraitSet traitSet,
+      List<RelNode> list,
+      RexNode rexNode,
+      Type type,
+      RelDataType relDataType,
+      Set<RelColumnMapping> set) {
+    return new BeamTableFunctionScanRel(
+        getCluster(), traitSet, list, rexNode, type, relDataType, columnMappings);
+  }
+
+  @Override
+  public PTransform<PCollectionList<Row>, PCollection<Row>> buildPTransform() {
+    return new Transform();
+  }
+
+  private class Transform extends PTransform<PCollectionList<Row>, PCollection<Row>> {
+
+    @Override
+    public PCollection<Row> expand(PCollectionList<Row> input) {
+      checkArgument(
+          input.size() == 1,
+          "Wrong number of inputs for %s: %s",
+          BeamTableFunctionScanRel.class.getSimpleName(),
+          input);
+      checkArgument(
+          ((RexCall) getCall()).getOperator().getName().equals("TUMBLE"),
+          "Only support TUMBLE table-valued function");
+      RexCall call = ((RexCall) getCall());
+      RexInputRef wmCol = (RexInputRef) call.getOperands().get(1);
+      PCollection<Row> upstream = input.get(0);
+      Schema outputSchema = CalciteUtils.toSchema(getRowType());
+      return upstream
+          .apply(
+              ParDo.of(
+                  new FixedWindowDoFn(
+                      FixedWindows.of(durationParameter(call.getOperands().get(2))),
+                      wmCol.getIndex(),
+                      outputSchema)))
+          .setRowSchema(outputSchema);
+    }
+  }
+
+  private Duration durationParameter(RexNode node) {
+    return Duration.millis(longValue(node));
+  }
+
+  private long longValue(RexNode operand) {
+    if (operand instanceof RexLiteral) {
+      return ((Number) RexLiteral.value(operand)).longValue();
+    } else {
+      throw new IllegalArgumentException(String.format("[%s] is not valid.", operand));
+    }
+  }
+
+  private static class FixedWindowDoFn extends DoFn<Row, Row> {
+    private int windowFieldIndex;
+    private FixedWindows windowFn;
+    private Schema outputSchema;
+
+    public FixedWindowDoFn(FixedWindows windowFn, int windowFieldIndex, Schema schema) {
+      this.windowFn = windowFn;
+      this.windowFieldIndex = windowFieldIndex;
+      this.outputSchema = schema;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      Row row = c.element();
+      IntervalWindow window = windowFn.assignWindow(row.getDateTime(windowFieldIndex).toInstant());
+      Row.Builder builder = Row.withSchema(outputSchema);
+      builder.addValues(row.getValues());
+      // TODO: add window_start and window_end;
```

Comment:
If this isn't to be resolved in this PR, would having an associated Jira be appropriate?

Or... is this TODO out of date? I see window start and end are used below.

Reply:
Ah it is out-of-date... Removed